.github/workflows/yocto-build-deploy.yml

```diff
@@ -828,6 +828,177 @@ jobs:
+          fi
+
+          sudo sysctl -w vm.mmap_min_addr=65536
+          sysctl vm.mmap_min_addr
+
+          mkdir -p "${SHARED_BUILD_DIR}"
+
+          # Source balena-lib to get helper functions
+          source "${automation_dir}/include/balena-lib.inc"
+
+          # Pull the helper image
+          if ! balena_lib_docker_pull_helper_image "${HELPER_IMAGE_REPO}" "" "yocto-build-env" helper_image_id; then
+            echo "Failed to pull helper image"
+            exit 1
+          fi
+
+          # Set up directories
+          DL_DIR="${SHARED_BUILD_DIR}/shared-downloads"
+          SSTATE_DIR="${SHARED_BUILD_DIR}/${MACHINE}/sstate"
+
+          # Step 1: Set up container
+          CONTAINER_NAME="verify-kernel-keys-$$"
+          docker run -d \
+            -v "${{ github.workspace }}":/work \
+            -v "${DL_DIR}":/yocto/shared-downloads \
+            -v "${SSTATE_DIR}":/yocto/shared-sstate \
+            -e BUILDER_UID="$(id -u)" \
+            -e VERBOSE="${VERBOSE}" \
+            -e BUILDER_GID="$(id -g)" \
+            --name "${CONTAINER_NAME}" \
+            --privileged \
+            "${helper_image_id}" \
+            sleep infinity || {
+            echo "Failed to start container"
+            exit 1
+          }
+
+          # Cleanup function
+          cleanup() {
+            docker stop "${CONTAINER_NAME}" 2>/dev/null || true
+            docker rm "${CONTAINER_NAME}" 2>/dev/null || true
+            # clean up temporary files
+            rm -f "$EXTRACTED_CERT" "$ORIGINAL_CERT_DER" "$ORIGINAL_CERT_PEM"
+          }
+          trap cleanup EXIT
+
+          # Step 2: Set up user (as root)
+          docker exec "${CONTAINER_NAME}" bash -c "set -e; \
+            groupadd -g \$BUILDER_GID builder 2>/dev/null || true; \
+            if ! cat /etc/group | grep docker > /dev/null; then groupadd docker; fi; \
+            useradd -m -u \$BUILDER_UID -g \$BUILDER_GID -G docker builder 2>/dev/null || true" || {
+            echo "Failed to set up container environment"
+            cleanup
+            exit 1
+          }
+
+          # Step 3: Extract bitbake environment as builder user
+          BITBAKE_ENV_OUTPUT=$(docker exec --user "$(id -u):$(id -g)" "${CONTAINER_NAME}" bash -c " \
+            INSTALL_DIR=/work; \
+            BUILD_DIR=build; \
+            if [ -f \${INSTALL_DIR}/layers/poky/oe-init-build-env ]; then \
+              source \${INSTALL_DIR}/layers/poky/oe-init-build-env \${INSTALL_DIR}/\${BUILD_DIR} > /dev/null 2>&1; \
+            else \
+              exit 1; \
+            fi; \
+            BITBAKE_ENV=\$(MACHINE=${MACHINE} bitbake -e virtual/kernel 2>&1); \
+            BITBAKE_EXIT=\$?; \
+            if [ \$BITBAKE_EXIT -eq 0 ] && [ -n \"\$BITBAKE_ENV\" ]; then \
+              echo \"\$BITBAKE_ENV\" | grep -E '^(B=|S=)'; \
+              exit 0; \
+            else \
+              exit 1; \
+            fi") || {
+            echo "Failed to extract bitbake environment"
```

Comment:
Would it be cleaner to `cat <<EOF > filename` to create these steps in a local script, then mount it in the container and run it?

It seems like there would be less escaping, and no security risk. You've managed to reduce it two 2 steps, but it still seems like overkill to run a detached container and exec into it multiple times.